File: mobile_app/lib/features/servicios/domain/servicio.dart
const suggestedServiceCategories = <String>[
  'residencial',
  'comercial',
  'industrial',
  'mantenimiento y emergencias',
];

String normalizeServiceCategory(String? value) {
  final normalized = value?.trim().toLowerCase() ?? '';
  return normalized;
}

String formatServiceCategoryLabel(String value) {
  final normalized = normalizeServiceCategory(value);
  if (normalized.isEmpty) {
    return '';
  }

  return normalized
      .split(' ')
      .where((part) => part.isNotEmpty)
      .map((part) => part[0].toUpperCase() + part.substring(1))
      .join(' ');
}

class Servicio {
  Servicio({
    required this.id,
    required this.codigo,
    required this.descripcion,
    required this.categoria,
    required this.unidad,
    required this.precioUnitario,
    required this.iva,
    required this.precioConIva,
    required this.observaciones,
    required this.activo,
  });

  final int id;
  final String codigo;
  final String descripcion;
  final String categoria;
  final String unidad;
  final String precioUnitario;
  final String iva;
  final String precioConIva;
  final String observaciones;
  final bool activo;

  factory Servicio.fromJson(Map<String, dynamic> json) {
    return Servicio(
      id: json['id'] as int,
      codigo: json['codigo']?.toString() ?? '',
      descripcion: json['descripcion']?.toString() ?? '',
      categoria: json['categoria']?.toString() ?? '',
      unidad: json['unidad']?.toString() ?? '',
      precioUnitario: json['precio_unitario']?.toString() ?? '0',
      iva: json['iva']?.toString() ?? '0',
      precioConIva: json['precio_con_iva']?.toString() ?? '0',
      observaciones: json['observaciones']?.toString() ?? '',
      activo: json['activo'] as bool? ?? false,
    );
  }
}

File: app/Models/Servicio.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;

class Servicio extends Model
{
    use HasFactory;

    public const CATEGORIAS_SUGERIDAS = [
        'residencial',
        'comercial',
        'industrial',
        'mantenimiento y emergencias',
    ];

    protected $fillable = [
        'categoria_servicio_id',
        'codigo',
        'nombre',
        'descripcion',
        'categoria',
        'unidad',
        'precio_unitario',
        'iva',
        'precio_con_iva',
        'observaciones',
        'precio_base',
        'activo',
    ];

    protected function casts(): array
    {
        return [
            'precio_unitario' => 'decimal:2',
            'iva' => 'decimal:2',
            'precio_con_iva' => 'decimal:2',
            'precio_base' => 'decimal:2',
            'activo' => 'boolean',
        ];
    }

    public function scopeBuscar(Builder $query, string $buscar): Builder
    {
        $buscar = trim($buscar);

        if ($buscar === '') {
            return $query;
        }

        return $query->where(function (Builder $subQuery) use ($buscar) {
            $subQuery->where('codigo', 'like', "%{$buscar}%")
                ->orWhere('descripcion', 'like', "%{$buscar}%")
                ->orWhere('categoria', 'like', "%{$buscar}%")
                ->orWhereHas('categoria', function (Builder $relationQuery) use ($buscar) {
                    $relationQuery->where('nombre', 'like', "%{$buscar}%");
                });
        });
    }

    public function scopeCategoria(Builder $query, ?string $categoria): Builder
    {
        $categoria = self::normalizeCategoria($categoria);

        if ($categoria === null) {
            return $query;
        }

        return $query->where(function (Builder $subQuery) use ($categoria) {
            $subQuery->whereRaw('LOWER(TRIM(categoria)) = ?', [$categoria])
                ->orWhereHas('categoria', function (Builder $relationQuery) use ($categoria) {
                    $relationQuery->whereRaw('LOWER(TRIM(nombre)) = ?', [$categoria]);
                });
        });
    }

    public static function normalizeCategoria(?string $categoria): ?string
    {
        if ($categoria === null) {
            return null;
        }

        $normalized = mb_strtolower(trim($categoria));

        if ($normalized === 'instalaciones') {
            return 'residencial';
        }

        return $normalized === '' ? null : $normalized;
    }

    public function categoria()
    {
        return $this->belongsTo(CategoriaServicio::class, 'categoria_servicio_id');
    }
}
